src/oom/oomd-util.c

```diff
@@ -0,0 +1,355 @@
+        uint64_t swap_threshold;
+        bool ret;
+
+        assert(threshold_percent <= 100);
+
+        swap_threshold = ctx.swap_total * ((uint64_t) threshold_percent) / 100;
+        ret = (ctx.swap_total - ctx.swap_used) < swap_threshold;
+        if (ret)
+                log_debug("Swap total (%"PRIu64") - used (%"PRIu64") is less than %u%%",
+                                ctx.swap_total, ctx.swap_used, threshold_percent);
+        return ret;
+}
+
+int compare_pgscan(const OomdCGroupContext *c1, const OomdCGroupContext *c2) {
+        assert(c1);
+        assert(c2);
+
+        if (c1->pgscan > c2->pgscan)
+                return 1;
+        else if (c1->pgscan < c2->pgscan)
+                return -1;
+        else
+                return 0;
+}
+
+int compare_swap_usage(const OomdCGroupContext *c1, const OomdCGroupContext *c2) {
+        assert(c1);
+        assert(c2);
+
+        if (c1->swap_usage > c2->swap_usage)
+                return 1;
+        else if (c1->swap_usage < c2->swap_usage)
+                return -1;
+        else
+                return 0;
+}
+
+int oomd_sort_cgroups(Hashmap *h, compare_func_t compare_func, const char *prefix, char ***ret) {
+        _cleanup_free_ char **sorted_cgroups = NULL;
+        OomdCGroupContext *c1, *c2;
+        Iterator i;
+        char *key;
+
+        assert(h);
+        assert(compare_func);
+        assert(ret);
+
+        HASHMAP_FOREACH_KEY(c1, key, h, i)
+                if (prefix == NULL || startswith(key, prefix)) {
+                        bool inserted = false;
+                        int r, j = 0;
+                        char **cgs;
+
+                        STRV_FOREACH(cgs, sorted_cgroups) {
+                                c2 = hashmap_get(h, *cgs);
+                                assert(c2);
+
+                                if (compare_func(c1, c2) > 0) {
+                                        r = strv_insert(&sorted_cgroups, j, key);
+                                        if (r < 0)
+                                                return log_oom();
+                                        inserted = true;
+                                        break;
+                                }
+
+                                j++;
+                        }
+
+                        if (!inserted) {
+                                r = strv_extend(&sorted_cgroups, key);
+                                if (r < 0)
+                                        return log_oom();
+                        }
+                }
+
+        *ret = TAKE_PTR(sorted_cgroups);
+        return 0;
+}
+
+int oomd_kill_by_pgscan(Hashmap *h, const char *prefix, bool dry_run) {
+        _cleanup_free_ char **sorted_cgroups = NULL;
+        char **cgs;
+        int r;
+
+        assert(h);
+        assert(prefix);
+
+        r = oomd_sort_cgroups(h, (compare_func_t) compare_pgscan, prefix, &sorted_cgroups);
```

Comment:
It's not entirely clear to me - but does this sort on all direct children of the monitored cgroup or all descendents?

You probably want this to be a "walk" of the tree, rooted at the monitored cgroup. E.g. enumerate all children, sort by the compare function. And recurse as necessary until you find a candidate to kill.

I would obey the cgroup's memory.oom.group setting (set via https://www.freedesktop.org/software/systemd/man/systemd.service.html#OOMPolicy=) - if it is 1, then we kill all descendant processes, otherwise keep recursing until we find a leaf cgroup (in which case we have no choice but to kill it regardless of OOMPolicy).

Reply:
Currently it sorts all descendants of the monitored cgroup.

I'll update this as suggested to look for leaf cgroups or memory.oom.group == 1.